dalle2_pytorch/train_diffusion_prior.py

```diff
@@ -0,0 +1,207 @@
+            emb_images = list(emb_images)
+            emb_text = list(emb_text)
+            emb_images[0] = torch.tensor(emb_images[0]).to(device)
+            emb_text[0] = torch.tensor(emb_text[0]).to(device)
+            model.eval()
+            loss = model(text_embed = emb_text[0],image_embed = emb_images[0])
+
+            # Log to wandb
+            if(isValidation):
+                wandb.log({"Validation mse": loss})
+            else:
+                wandb.log({"Test mse": loss})
+
+    # Saving State Dict
+    torch.save(model.state_dict(), 'saved_model.pth')
+
+def train(image_embed_dim,
+          image_embed_url,
+          text_embed_url,
+          batch_size,
+          train_percent,
+          val_percent,
+          test_percent,
+          num_epochs,
+          dp_loss_type,
+          clip,
+          dp_condition_on_text_encodings,
+          dp_timesteps,
+          dp_cond_drop_prob,
+          dpn_depth,
+          dpn_dim_head,
+          dpn_heads,
+          device,
+          learning_rate=0.01):
+
+
+    # DiffusionPriorNetwork 
+    prior_network = DiffusionPriorNetwork( 
+            dim = image_embed_dim, 
+            depth = dpn_depth, 
+            dim_head = dpn_dim_head, 
+            heads = dpn_heads).to(device)
+    
+    # DiffusionPrior with text embeddings and image embeddings pre-computed
+    diffusion_prior = DiffusionPrior( 
+            net = prior_network, 
+            clip = clip, 
+            image_embed_dim = image_embed_dim, 
+            timesteps = dp_timesteps,
+            cond_drop_prob = dp_cond_drop_prob, 
+            loss_type = dp_loss_type, 
+            condition_on_text_encodings = dp_condition_on_text_encodings).to(device)
+
+    # Get image and text embeddings from the servers
+    print("==============Downloading embeddings - image and text====================")
+    image_reader = EmbeddingReader(embeddings_folder=image_embed_url, file_format="npy")
+    text_reader = EmbeddingReader(embeddings_folder=text_embed_url, file_format="npy")
+    num_data_points = text_reader.count
+
+    ### Training code ###
+    optimizer = get_optimizer(diffusion_prior.parameters())
+    epochs = num_epochs
+
+    for _ in range(epochs):
+        train_set_size = int(train_percent*num_data_points)
+        for emb_images,emb_text in zip(image_reader(batch_size=batch_size, start=0, end=train_set_size),
+                text_reader(batch_size=batch_size, start=0, end=train_set_size)):
+            emb_images = list(emb_images)
+            emb_text = list(emb_text)
+            emb_images[0] = torch.tensor(emb_images[0]).to(device)
+            emb_text[0] = torch.tensor(emb_text[0]).to(device)
+            optimizer.zero_grad()
+            loss = diffusion_prior(text_embed = emb_text[0],image_embed = emb_images[0])
+            loss.backward()
+            # Log to wandb
+            wandb.log({"Training (mse)": loss})
```

Comment:
Would be great to report here as well:
* Number of steps (just add a step = 0 at the top and you do step= step+1 in the inner loop)
* Sample/s (you simply need a t = time.time() variable at the top and do sample_per_sec = step / (time.time() - t)